### ISSUE TITLE
Lizmap 4 development environnement

### ISSUE THREAD

**Issue by nboisteault** [3liz/lizmap-rfc#1]
# Lizmap RFC: Lizmap 4 development environnement

**Date** 2019/07/30

**Author** Nicolas Boisteault (@nboisteault )

**Contact** info@3liz.com

**maintainer** @nboisteault

**Version** Lizmap-Web-Client 4.0

**Developing branch** https://github.com/3liz/lizmap-web-client/tree/dev_lizmap4

# Summary

**This RFC is about all the pieces that will compose the Lizmap 4 development environnement.**

## Proposed Solution

- OpenLayers 6+
- [Unidirectional User Interface Architecture](https://github.com/3liz/lizmap-rfc/issues/3)
- ECMAScript 6
- Web components
- Webpack
- Babel.js + polyfills for old browsers
- [Fancytree](https://github.com/mar10/fancytree) for layer's tree.
- JSDoc for Javascript documentation.
- ESLint

### Affected Files

All previous Javascript scripts will be dropped (front-end).
Many parts of the UI (templates) previously handled  in the PHP back-end will be handled by web components (front-end).

## Browser compatibility

All modern browsers, MS Edge 18+

## Backwards Compatibility

- **Lizmap configuration files made by the Lizmap plugin (*.cfg) will have to be converted to a new version**

- **Lizmap 4 will be compatible with QGIS Server >= 3.4**

## Issue Tracking ID(s)

https://github.com/3liz/lizmap-web-client/issues/925

## Votes

(required)

**Comment by rldhont** [comment]
Thanks @nboisteault.

Which EcmaScript version ?
Which browser compatibility ?

**Comment by Raruto** [comment]
Which version of openlayers?

I think it would be more correct to talk about MS Edge 76+ (chromium-based version), since:
- there will be no future versions after IE 11, 
- web components will supported only in future MS Edge browsers

**PS**: unless babel isn't already able to transiple them...

**Comment by nboisteault** [comment]
@Raruto We'll use babel to transpile cutting edge ECMAScript to < ECMAScript 6 for browser compatibility with old ones.

**Comment by Raruto** [comment]
Yep, but since web compontents are also a HTML 5 spec, you will probably still need a [polyfill](https://github.com/webcomponents/polyfills) library (read [here](https://github.com/github/babel-plugin-transform-custom-element-classes#compatibility) for more info)

**Comment by nboisteault** [comment]
@Raruto yes I have edited RFC to add the use of polyfills. Thanks.

**Comment by rldhont** [comment]
@nboisteault I have removed FLUX mention and replaced by Unidirectional User Interface Architecture.
